### PR TITLE
[1/N] pass the complete config from engine to executor

### PR DESCRIPTION
**Pull request by youkaichao**

### CODE REVIEW

**Comment by youkaichao** [comment]
the path is:

Engine --> Executor --> Worker --> ModelRunner --> Model

all of them will get the complete engine config object

File: vllm/engine/llm_engine.py [hunk 1]
```diff
@@ -222,24 +219,29 @@ def validate_outputs(
...
+
+        model_config = self.model_config = vllm_config.model_config
+        cache_config = self.cache_config = vllm_config.cache_config
+        lora_config = self.lora_config = vllm_config.lora_config
+        parallel_config = self.parallel_config = vllm_config.parallel_config
+        scheduler_config = self.scheduler_config = vllm_config.scheduler_config
+        device_config = self.device_config = vllm_config.device_config
+        speculative_config = self.speculative_config = vllm_config.speculative_config  # noqa
+        load_config = self.load_config = vllm_config.load_config
+        decoding_config = self.decoding_config = vllm_config.decoding_config or DecodingConfig(  # noqa
```
Comment (russellb) [inline vllm/engine/llm_engine.py:239]: I assume the extra local variables here are to minimize the code changes necessary in this step?

If so, maybe add a "TODO" comment to come back and remove them eventually?

File: vllm/v1/engine/llm_engine.py [hunk 2]
```diff
@@ -35,24 +32,29 @@ class LLMEngine:
...
+        lora_config = self.lora_config = vllm_config.lora_config
+        parallel_config = self.parallel_config = vllm_config.parallel_config
+        scheduler_config = self.scheduler_config = vllm_config.scheduler_config
+        device_config = self.device_config = vllm_config.device_config
+        speculative_config = self.speculative_config = vllm_config.speculative_config  # noqa
+        load_config = self.load_config = vllm_config.load_config
+        decoding_config = self.decoding_config = vllm_config.decoding_config or DecodingConfig(  # noqa
+        )
+        prompt_adapter_config = self.prompt_adapter_config = vllm_config.prompt_adapter_config  # noqa
+        observability_config = self.observability_config = vllm_config.observability_config or ObservabilityConfig(  # noqa
```
Comment (russellb) [inline vllm/v1/engine/llm_engine.py:55]: and the same TODO comment here?

**Review by russellb — COMMENTED** [review]
This looks like a nice step in refactoring. I had one suggestion about leaving a TODO note in the code.

**Comment by robertgshaw2-neuralmagic** [comment]
+1 this is a nice change

File: vllm/engine/llm_engine.py (hunk 1 above)
Comment (youkaichao) [reply vllm/engine/llm_engine.py:239]: yeah, exactly, I do this for minimizing code change in this pr.

Comment (robertgshaw2-neuralmagic) [reply vllm/engine/llm_engine.py:239]: Unfortunately, these are local variables are primarily used in unit testing

**Comment by youkaichao** [comment]
> +1 this is a nice change

thanks for your appreciation! the goal is:

if we have a config field that is only needed by one component, we don't need to change the function signature for all these classes.

(context: I'm going to add a compilation config for `torch.compile` , and it is only needed by the final model.)

**Comment by youkaichao** [comment]
each component can still save their local copy of configs like `self.model_config = vllm_config.model_config` so that they don't need to always start with `self.vllm_config` . but the point is that all of them will have `self.vllm_config` and pass it as a whole to others.

File: vllm/engine/llm_engine.py (hunk 1 above)
Comment (youkaichao) [reply vllm/engine/llm_engine.py:239]: we can clean up a little bit inside this function though.

File: vllm/v1/engine/llm_engine.py (hunk 2 above)
Comment (youkaichao) [reply vllm/v1/engine/llm_engine.py:56]: added in [6c1773a](https://github.com/vllm-project/vllm/pull/9933/commits/6c1773a17fe1c3cde83f3094d3ef479b0ef415e8)

**Review by robertgshaw2-redhat — APPROVED** [review]

**Comment by comaniac** [comment]
After this PR, the following command hangs during startup:

```
VLLM_USE_V1=1 vllm serve neuralmagic/Meta-Llama-3-8B-Instruct-FP8 --disable-log-requests
```

Log:
```
(base) ray@ip-10-0-54-152:~/default/vllm$ VLLM_USE_V1=1 vllm serve neuralmagic/Meta-Llama-3-8B-Instruct-FP8 --disable-log-requests
/home/ray/anaconda3/lib/python3.11/site-packages/vllm/connections.py:8: RuntimeWarning: Failed to read commit hash:
No module named 'vllm._version'
  from vllm.version import __version__ as VLLM_VERSION
INFO 11-03 22:37:01 api_server.py:551] vLLM API server version dev
INFO 11-03 22:37:01 api_server.py:552] args: Namespace(subparser='serve', model_tag='neuralmagic/Meta-Llama-3-8B-Instruct-FP8', config='', host=None, port=8000, uvicorn_log_level='info', allow_credentials=False, allowed_origins=['*'], allowed_methods=['*'], allowed_headers=['*'], api_key=None, lora_modules=None, prompt_adapters=None, chat_template=None, response_role='assistant', ssl_keyfile=None, ssl_certfile=None, ssl_ca_certs=None, ssl_cert_reqs=0, root_path=None, middleware=[], return_tokens_as_token_ids=False, disable_frontend_multiprocessing=False, enable_auto_tool_choice=False, tool_call_parser=None, tool_parser_plugin='', model='neuralmagic/Meta-Llama-3-8B-Instruct-FP8', task='auto', tokenizer=None, skip_tokenizer_init=False, revision=None, code_revision=None, tokenizer_revision=None, tokenizer_mode='auto', chat_template_text_format='string', trust_remote_code=False, download_dir=None, load_format='auto', config_format=<ConfigFormat.AUTO: 'auto'>, dtype='auto', kv_cache_dtype='auto', quantization_param_path=None, max_model_len=None, guided_decoding_backend='outlines', distributed_executor_backend=None, worker_use_ray=False, pipeline_parallel_size=1, tensor_parallel_size=1, max_parallel_loading_workers=None, ray_workers_use_nsight=False, block_size=16, enable_prefix_caching=False, disable_sliding_window=False, use_v2_block_manager=False, num_lookahead_slots=0, seed=0, swap_space=4, cpu_offload_gb=0, gpu_memory_utilization=0.9, num_gpu_blocks_override=None, max_num_batched_tokens=None, max_num_seqs=256, max_logprobs=20, disable_log_stats=False, quantization=None, rope_scaling=None, rope_theta=None, enforce_eager=False, max_seq_len_to_capture=8192, disable_custom_all_reduce=False, tokenizer_pool_size=0, tokenizer_pool_type='ray', tokenizer_pool_extra_config=None, limit_mm_per_prompt=None, mm_processor_kwargs=None, enable_lora=False, max_loras=1, max_lora_rank=16, lora_extra_vocab_size=256, lora_dtype='auto', long_lora_scaling_factors=None, max_cpu_loras=None, fully_sharded_loras=False, enable_prompt_adapter=False, max_prompt_adapters=1, max_prompt_adapter_token=0, device='auto', num_scheduler_steps=1, multi_step_stream_outputs=True, scheduler_delay_factor=0.0, enable_chunked_prefill=None, speculative_model=None, speculative_model_quantization=None, num_speculative_tokens=None, speculative_disable_mqa_scorer=False, speculative_draft_tensor_parallel_size=None, speculative_max_model_len=None, speculative_disable_by_batch_size=None, ngram_prompt_lookup_max=None, ngram_prompt_lookup_min=None, spec_decoding_acceptance_method='rejection_sampler', typical_acceptance_sampler_posterior_threshold=None, typical_acceptance_sampler_posterior_alpha=None, disable_logprobs_during_spec_decoding=None, model_loader_extra_config=None, ignore_patterns=[], preemption_mode=None, served_model_name=None, qlora_adapter_name_or_path=None, otlp_traces_endpoint=None, collect_detailed_traces=None, disable_async_output_proc=False, override_neuron_config=None, scheduling_policy='fcfs', pooling_type=None, pooling_norm=None, pooling_softmax=None, pooling_step_tag_id=None, pooling_returned_token_ids=None, disable_log_requests=True, max_log_len=None, disable_fastapi_docs=False, dispatch_function=<function serve at 0x7ef3e97a8540>)
INFO 11-03 22:37:01 api_server.py:166] Multiprocessing frontend to use ipc:///tmp/6bd30e7c-ebe7-47a7-9a63-b75add142a3b for IPC Path.
INFO 11-03 22:37:01 api_server.py:181] Started engine process with PID 14128
/home/ray/anaconda3/lib/python3.11/site-packages/vllm/connections.py:8: RuntimeWarning: Failed to read commit hash:
No module named 'vllm._version'
  from vllm.version import __version__ as VLLM_VERSION
INFO 11-03 22:37:06 config.py:323] This model supports multiple tasks: {'embedding', 'generate'}. Defaulting to 'generate'.
WARNING 11-03 22:37:06 arg_utils.py:1103] [DEPRECATED] Block manager v1 has been removed, and setting --use-v2-block-manager to True or False has no effect on vLLM behavior. Please remove --use-v2-block-manager in your engine argument. If your use case is not supported by SelfAttnBlockSpaceManager (i.e. block manager v2), please file an issue with detailed information.
INFO 11-03 22:37:10 config.py:323] This model supports multiple tasks: {'generate', 'embedding'}. Defaulting to 'generate'.
WARNING 11-03 22:37:10 arg_utils.py:1103] [DEPRECATED] Block manager v1 has been removed, and setting --use-v2-block-manager to True or False has no effect on vLLM behavior. Please remove --use-v2-block-manager in your engine argument. If your use case is not supported by SelfAttnBlockSpaceManager (i.e. block manager v2), please file an issue with detailed information.
INFO 11-03 22:37:10 llm_engine.py:69] Initializing an LLM engine (vdev) with config: model='neuralmagic/Meta-Llama-3-8B-Instruct-FP8', speculative_config=None, tokenizer='neuralmagic/Meta-Llama-3-8B-Instruct-FP8', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, rope_scaling=None, rope_theta=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.bfloat16, max_seq_len=8192, download_dir=None, load_format=LoadFormat.AUTO, tensor_parallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=fp8, enforce_eager=False, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='outlines'), observability_config=ObservabilityConfig(otlp_traces_endpoint=None, collect_model_forward_time=False, collect_model_execute_time=False), seed=0, served_model_name=neuralmagic/Meta-Llama-3-8B-Instruct-FP8, num_scheduler_steps=1, enable_prefix_caching=False, use_async_output_proc=True, mm_processor_kwargs=None)
/home/ray/anaconda3/lib/python3.11/site-packages/vllm/connections.py:8: RuntimeWarning: Failed to read commit hash:
No module named 'vllm._version'
  from vllm.version import __version__ as VLLM_VERSION
# Hanging for minutes
```

**Pull request merged**